### PR TITLE
Task ys/j51 edit podcast functionality

### PR DESCRIPTION
**Pull request by yaroslav207**

### CODE REVIEW

File: frontend/src/store/configurate-podcast/reducer.ts [hunk 1]
```diff
@@ -1,16 +1,18 @@
...
 type State = {
   dataStatus: DataStatus;
   podcast: Podcast | null;
+  editablePodcast: Podcast | null
 };
 
 const initialState: State = {
   dataStatus: DataStatus.IDLE,
   podcast: null,
+  editablePodcast: null,
```
Comment (ghost) [inline frontend/src/store/configurate-podcast/reducer.ts]:
```suggestion
};

const initialState: State = {
  dataStatus: DataStatus.IDLE,
  podcast: null,
```

File: frontend/src/services/podcast-api/podcast-api.service.ts [hunk 2]
```diff
@@ -49,6 +49,17 @@ class PodcastApi {
       },
     );
   }
+
+  public edit(id: number, payload: PodcastEditPayload): Promise<Podcast> {
```
Comment (ghost) [inline frontend/src/services/podcast-api/podcast-api.service.ts]:
```suggestion
  public edit(payload: PodcastEditPayload): Promise<Podcast> {
```

File: shared/src/common/enums/podcast/podcast-edit-payload-key.enum.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+enum PodcastEditPayloadKey {
```
Comment (ghost) [inline shared/src/common/enums/podcast/podcast-edit-payload-key.enum.ts]: I think that we don't need and can use `Podcast` type instead

File: frontend/src/common/types/podcast/podcast-edit-action.type.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { PodcastEditPayload } from './podcast';
```
Comment (ghost) [inline frontend/src/common/types/podcast/podcast-edit-action.type.ts]: not sure that we need it, we can use regular Podcast type

File: frontend/src/components/configurate-podcast/configurate-podcast.tsx [hunk 5]
```diff
@@ -1,26 +1,46 @@
...
+
   const isEdit = Boolean(id);
 
   const handleCreatePodcast = (payload: PodcastCreatePayload): void => {
     dispatch(configuratePodcastActions.create(payload));
   };
 
+  const handleEditPodcast = (payload: PodcastEditPayload): void => {
+    dispatch(configuratePodcastActions.edit( { id: Number(id), payload } ));
+  };
```
Comment (ghost) [inline frontend/src/components/configurate-podcast/configurate-podcast.tsx]:
```suggestion
  const handleFormSubmit = (payload: PodcastEditPayload): void => {
    const isEdit = Boolean(payload);
    isEdit ? dispatch(actions.edit(payload)) :  dispatch(actions.create(payload))

  };
```

File: frontend/src/services/podcast-api/podcast-api.service.ts [hunk 6]
```diff
@@ -49,6 +49,17 @@ class PodcastApi {
...
+    return this.#http.load(
+      `${this.#apiPrefix}${ApiPath.PODCASTS}${PodcastsApiPath.ROOT}${id}`,
```
Comment (ghost) [inline frontend/src/services/podcast-api/podcast-api.service.ts]:
```suggestion
      `${this.#apiPrefix}${ApiPath.PODCASTS}${PodcastsApiPath.ROOT}${payload.id}`,
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: frontend/src/components/configurate-podcast/configurate-podcast.tsx [hunk 7]
```diff
@@ -9,18 +9,46 @@ const ConfiguratePodcast: React.FC = () => {
...
+    if (isEdit) {
+      dispatch(configuratePodcastActions.loadPodcast(Number(id)));
+    }
+  }, []);
+
+  useEffect(() => {
+    if (isEdit) {
+      dispatch(configuratePodcastActions.loadPodcast(Number(id)));
+    }
+  }, []);
```
Comment (yevhenprybytok) [inline frontend/src/components/configurate-podcast/configurate-podcast.tsx:34]: Same code duplicated

File: frontend/src/components/configurate-podcast/configurate-podcast.tsx [hunk 8]
```diff
@@ -9,18 +9,46 @@ const ConfiguratePodcast: React.FC = () => {
...
+  const { podcast } = useAppSelector(({ configuratePodcast }) => ({
+    podcast: configuratePodcast.podcast,
+  }));
+
   const isEdit = Boolean(id);
 
-  const handleSubmit = (payload: PodcastFormPayload): void => {
-    dispatch(configuratePodcastActions.create(payload));
+  const handleFormSubmit = (payload: PodcastFormPayload): void => {
+    isEdit ? dispatch(configuratePodcastActions.edit(payload)) : dispatch(configuratePodcastActions.create(payload));
```
Comment (yevhenprybytok) [inline frontend/src/components/configurate-podcast/configurate-podcast.tsx]: Create separate methods for those actions

File: frontend/src/components/configurate-podcast/configurate-podcast.tsx [hunk 9]
```diff
@@ -9,18 +9,46 @@ const ConfiguratePodcast: React.FC = () => {
...
         {isEdit ? 'Edit' : 'Create'} Podcast {id ?? ''}
       </h2>
-      <ConfiguratePodcastForm onSubmit={handleSubmit} />;
+
+      {(isEdit)
+        ? podcast && <ConfiguratePodcastForm
+          onSubmit={handleFormSubmit}
+          payload={mapPodcastToFormPayload(podcast)}/>
+        : <ConfiguratePodcastForm onSubmit={handleFormSubmit}/>
+      }
```
Comment (yevhenprybytok) [inline frontend/src/components/configurate-podcast/configurate-podcast.tsx]:
```suggestion
      <ConfiguratePodcastForm
          onSubmit={handleFormSubmit}
          payload={isEdit ? mappedPodcast : null}
       />
```
mappedPodcast - is the result of calling mapPodcastToFormPayload

File: frontend/src/store/configurate-podcast/actions.ts [hunk 10]
```diff
@@ -24,4 +24,29 @@ const create = createAsyncThunk<Podcast, PodcastFormPayload, AsyncThunkConfig>(
   },
 );
 
-export { create };
+const edit = createAsyncThunk<Podcast | null, PodcastFormPayload, AsyncThunkConfig>
+(ActionType.EDIT_PODCAST, async ( podcastPayload, { getState, extra }) => {
+  const { configuratePodcast } = getState();
+  const { podcastApi } = extra;
+  if(configuratePodcast.podcast){
```
Comment (yevhenprybytok) [inline frontend/src/store/configurate-podcast/actions.ts]:
```suggestion

  if (configuratePodcast.podcast) {
```

**Review by yevhenprybytok — CHANGES_REQUESTED** [review]

File: backend/src/services/podcast/podcast.service.ts [hunk 11]
```diff
@@ -77,6 +79,36 @@ class Podcast {
...
+    imageDataUrl,
+  }: PodcastEditPayload): Promise<TPodcast> {
+    const updatePodcast: PodcastEditDTOPayload = {
+      name,
+      description,
+      imageId: imageId,
+    };
+
+    if (imageDataUrl) {
+      const { url, publicId } = await this.#fileStorage.upload({
```
Comment (ghost) [inline backend/src/services/podcast/podcast.service.ts:99]: we need to delete the old image

File: frontend/src/components/configurate-podcast/components/configurate-podcast-form/configurate-podcast-form.tsx [hunk 12]
```diff
@@ -15,6 +15,7 @@ import { DEFAULT_PODCAST_PAYLOAD } from './common/constants';
 
 type Props = {
   onSubmit: (payload: PodcastFormPayload) => void;
+  payload: PodcastFormPayload | undefined
```
Comment (ghost) [inline frontend/src/components/configurate-podcast/components/configurate-podcast-form/configurate-podcast-form.tsx]:
```suggestion
  payload?: PodcastFormPayload
```

File: frontend/src/components/configurate-podcast/configurate-podcast.tsx [hunk 13]
```diff
@@ -9,18 +9,39 @@ const ConfiguratePodcast: React.FC = () => {
...
+  const { podcast, dataStatus } = useAppSelector(({ configuratePodcast }) => ({
+    podcast: configuratePodcast.podcast,
+    dataStatus: configuratePodcast.dataStatus,
+  }));
+
   const isEdit = Boolean(id);
 
-  const handleSubmit = (payload: PodcastFormPayload): void => {
-    dispatch(configuratePodcastActions.create(payload));
+  const mapPodcastToFormPayload = (podcast: Podcast): PodcastFormPayload => ({
```
Comment (ghost) [inline frontend/src/components/configurate-podcast/configurate-podcast.tsx]: move it to `frontend/src/components/configurate-podcast/helpers`

File: frontend/src/components/configurate-podcast/configurate-podcast.tsx [hunk 14]
```diff
@@ -9,18 +9,39 @@ const ConfiguratePodcast: React.FC = () => {
...
+    }
+  }, []);
+
   return (
     <div className={styles.podcast}>
       <h2>
         {isEdit ? 'Edit' : 'Create'} Podcast {id ?? ''}
       </h2>
-      <ConfiguratePodcastForm onSubmit={handleSubmit} />;
+      {(!(dataStatus === 'pending')) && <ConfiguratePodcastForm onSubmit={handleFormSubmit} payload={mapPodcast}/>}
```
Comment (ghost) [inline frontend/src/components/configurate-podcast/configurate-podcast.tsx]: return loading

File: frontend/src/services/podcast-api/podcast-api.service.ts [hunk 15]
```diff
@@ -6,6 +6,7 @@ import {
 } from 'common/enums/enums';
 import { Podcast, PodcastCreatePayload } from 'common/types/types';
 import { Http } from 'services/http/http.service';
+import { PodcastEditPayload } from 'common/types/types';
```
Comment (ghost) [inline frontend/src/services/podcast-api/podcast-api.service.ts]: move above

File: frontend/src/store/configurate-podcast/actions.ts [hunk 16]
```diff
@@ -24,4 +24,32 @@ const create = createAsyncThunk<Podcast, PodcastFormPayload, AsyncThunkConfig>(
...
 );
 
-export { create };
+const edit = createAsyncThunk<Podcast | null, PodcastFormPayload, AsyncThunkConfig>
+(ActionType.EDIT_PODCAST, async ( podcastPayload, { getState, extra }) => {
+  const { podcastApi } = extra;
+  const { configuratePodcast } = getState();
+  const [file] = podcastPayload.image ?? [];
+  if (configuratePodcast.podcast) {
+    const payloadEditPodcast = {
```
Comment (ZubkoWhatislove) [inline frontend/src/store/configurate-podcast/actions.ts:33]: move it to function call

File: frontend/src/store/configurate-podcast/actions.ts [hunk 17]
```diff
@@ -24,4 +24,32 @@ const create = createAsyncThunk<Podcast, PodcastFormPayload, AsyncThunkConfig>(
...
+      name: podcastPayload.name,
+      description: podcastPayload.description,
+      imageDataUrl: file ? await getDataUrl(file) : null,
+    };
+
+    const podcast = await podcastApi.edit(payloadEditPodcast);
+
+    return podcast;
+  }
+  return null;
```
Comment (ghost) [inline frontend/src/store/configurate-podcast/actions.ts]:
```suggestion
```

File: frontend/src/store/configurate-podcast/actions.ts [hunk 18]
```diff
@@ -24,4 +24,32 @@ const create = createAsyncThunk<Podcast, PodcastFormPayload, AsyncThunkConfig>(
   },
 );
 
-export { create };
+const edit = createAsyncThunk<Podcast | null, PodcastFormPayload, AsyncThunkConfig>
```
Comment (ghost) [inline frontend/src/store/configurate-podcast/actions.ts]:
```suggestion
const edit = createAsyncThunk<Podcast, PodcastFormPayload, AsyncThunkConfig>
```

File: shared/src/common/enums/podcast/podcast-edit-payload-key.enum.ts [hunk 19]
```diff
@@ -0,0 +1,12 @@
...
```
Comment (ghost) [inline shared/src/common/enums/podcast/podcast-edit-payload-key.enum.ts]: remove it and use `podcast-payload-key.enum.ts `

File: shared/src/common/enums/podcast/podcast.ts [hunk 20]
```diff
@@ -1 +1,2 @@
+export { PodcastEditPayloadKey } from './podcast-edit-payload-key.enum';
```
Comment (ghost) [inline shared/src/common/enums/podcast/podcast.ts]:
```suggestion
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: backend/src/services/podcast/podcast.service.ts (hunk 11 above)
Comment (ghost) [reply backend/src/services/podcast/podcast.service.ts:99]: don't fixed

File: frontend/src/components/configurate-podcast/configurate-podcast.tsx [hunk 21]
```diff
@@ -1,26 +1,45 @@
...
+    }
+  }, []);
+
   return (
     <div className={styles.podcast}>
       <h2>
         {isEdit ? 'Edit' : 'Create'} Podcast {id ?? ''}
       </h2>
-      <ConfiguratePodcastForm onSubmit={handleSubmit} />;
+      {(dataStatus === 'pending')
```
Comment (ghost) [inline frontend/src/components/configurate-podcast/configurate-podcast.tsx]:
```suggestion
      {(dataStatus === DataStatus.PENDING)
```

File: frontend/src/helpers/file/parse-file/parse-file.ts [hunk 22]
```diff
@@ -0,0 +1,6 @@
+const parseFile = (payload: FileList | null): File => {
```
Comment (ghost) [inline frontend/src/helpers/file/parse-file/parse-file.ts]: don't forget to change the filename
```suggestion
const getFileFromFileList = (payload: FileList | null): File => {
```

File: frontend/src/helpers/podcast/map-podcast-to-form-payload/map-podcast-to-form-payload.ts [hunk 23]
```diff
@@ -0,0 +1,9 @@
+import { Podcast, PodcastFormPayload } from 'common/types/types';
```
Comment (ghost) [inline frontend/src/helpers/podcast/map-podcast-to-form-payload/map-podcast-to-form-payload.ts]: move it to helper to place where you use it

File: frontend/src/services/podcast-api/podcast-api.service.ts [hunk 24]
```diff
@@ -49,6 +49,17 @@ class PodcastApi {
       },
     );
   }
+
+  public edit(payload: PodcastEditPayload): Promise<Podcast> {
```
Comment (ghost) [inline frontend/src/services/podcast-api/podcast-api.service.ts]:
```suggestion
  public update(payload: PodcastEditPayload): Promise<Podcast> {
```

File: frontend/src/store/configurate-podcast/actions.ts (hunk 16 above)
Comment (ZubkoWhatislove) [reply frontend/src/store/configurate-podcast/actions.ts:34]: don't fixed

**Review by ghost — CHANGES_REQUESTED** [review]

File: backend/src/data/repositories/image/image.repository.ts [hunk 25]
```diff
@@ -15,6 +15,10 @@ class Image {
   public create(payload: ImageCreatePayload): Promise<TImage> {
     return this.#ImageModel.query().insert(payload);
   }
+
+  public delete(payload: number): Promise<number> {
+    return this.#ImageModel.query().deleteById(payload);
```
Comment (ghost) [inline backend/src/data/repositories/image/image.repository.ts]:
```suggestion
  public delete(id: number): Promise<number> {
    return this.#ImageModel.query().deleteById(id);
```

File: frontend/src/components/configurate-podcast/configurate-podcast.tsx [hunk 26]
```diff
@@ -1,26 +1,47 @@
-import { useDispatch, useParams } from 'hooks/hooks';
+import { useDispatch, useParams, useEffect, useAppSelector } from 'hooks/hooks';
 import { configuratePodcast as configuratePodcastActions } from 'store/actions';
 import { PodcastFormPayload } from 'common/types/types';
 import { PageParams } from './common/types/types';
+
```
Comment (ghost) [inline frontend/src/components/configurate-podcast/configurate-podcast.tsx]:
```suggestion
```

File: frontend/src/store/configurate-podcast/actions.ts [hunk 27]
```diff
@@ -24,4 +24,30 @@ const create = createAsyncThunk<Podcast, PodcastFormPayload, AsyncThunkConfig>(
...
+    const updatePodcast = <Podcast>configuratePodcast.podcast;
+    const payloadEditPodcast = {
+      ...updatePodcast,
+      name: podcastPayload.name,
+      description: podcastPayload.description,
+      imageDataUrl: file ? await getDataUrl(file) : null,
+    };
+    const podcast = await podcastApi.update(payloadEditPodcast);
+
+    return podcast;
```
Comment (ghost) [inline frontend/src/store/configurate-podcast/actions.ts]:
```suggestion
   return podcastApi.update({
          ...updatePodcast,
      name: podcastPayload.name,
      description: podcastPayload.description,
      imageDataUrl: file ? await getDataUrl(file) : null,
    });
```

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**